### PR TITLE
Document SID bounds API

### PR DESCRIPTION
**Pull request by anstaf**
- `get_lower_bound`/`get_upper_bound` helpers are added to the sid concept
- SID boundaries API is documented

### CODE REVIEW

File: include/gridtools/sid/concept.hpp [hunk 1]
```diff
@@ -89,14 +94,30 @@
...
+ *     `offset < at_key<a>(lower_bounds)`.
+ *    The same logic is applied to the `UpperBounds`: the shift is illegal if `offset >= at_key<a>(upper_bounds)`.
+ *    Note that the absence of the key in the bounds means nothing.
+ *    Note also that there is no contract that `sid_shift` is always legal with any `offset` within the boundaries.
+ *
+ *    In some sense SID boundaries is a "best effort" API. The users of the SID concepts can use it to validate the
+ *    data boundaries against the computation area. The style of that validation should be that: if there are boundaries
+ *    along that dimension, we check, otherwise everything is OK. The SID modelers are free to skip boundaries
+ *    definitions. There are strong use cases where that skipping makes sense. For example blocking. After the blocking
+ *    the valid iteration area can be not rectangular (even the original SID has the rectangular one). Even we would
```
Comment (egparedes) [inline include/gridtools/sid/concept.hpp:109]: If I understand correctly this sentence:

    "...  the valid iteration area can be not rectangular (even the original SID has the rectangular one)"

I think it'd be more clear like:

    "...  the valid iteration area can be not rectangular (even when the original SID is rectangular)" ?

Comment (anstaf) [reply include/gridtools/sid/concept.hpp]: done

File: include/gridtools/sid/concept.hpp [hunk 2]
```diff
@@ -51,8 +54,10 @@
  *     - there is `PtrHolder operator+(PtrHolder, PtrDiff)` defined and `(holder + diff)()` should return the same as
  *       `holder() + diff`
  *     - decayed `Strides` is a hymap
+ *     - decayed  `LowerBounds` and decayed `UpperBounds` are hymaps. The values of the those hymaps are of integral
```
Comment (havogt) [inline include/gridtools/sid/concept.hpp:57]:
```suggestion
 *     - decayed `LowerBounds` and decayed `UpperBounds` are hymaps. The values of the those hymaps are of integral
```

File: include/gridtools/sid/concept.hpp [hunk 3]
```diff
@@ -89,14 +94,30 @@
  *    For concept implementors it means that the inferred from the `SID` types can delegate the ownership handling
  *    to `SID`. It is legal for example the `Strides` can be implemented as a reference (or a tuple of references)
  *
- *    TODO(anstaf): formal semantic definition is not complete.
+ *    `LowerBounds` and `UpperBounds` define the boundaries of the iteration space. They model `hymap` concept as well
+ *    as `Strides` does. However the keys not necessarily must be the same. If some key (`a`) exists both in `Strides`
+ *    and `LowerBounds` the call of `sid_shift(origin, at_key<a>(strides), offset)` will be illegal where
+ *     `offset < at_key<a>(lower_bounds)`.
+ *    The same logic is applied to the `UpperBounds`: the shift is illegal if `offset >= at_key<a>(upper_bounds)`.
+ *    Note that the absence of the key in the bounds means nothing.
```
Comment (havogt) [inline include/gridtools/sid/concept.hpp:102]:
```suggestion
 *    Note that the absence of the key in the bounds means there is no check.
```

**Comment by havogt** [comment]
Only minor comment

**Comment by anstaf** [comment]
launch jenkins

**Pull request merged**